R/as.data.table.R

```diff
@@ -245,13 +245,14 @@ as.data.table.data.frame = function(x, keep.rownames=FALSE, key=NULL, ...) {
   ans
 }
 
-as.data.table.data.table = function(x, ...) {
+as.data.table.data.table = function(x, keep.rownames, key=NULL, ...) {
```

Comment:
```suggestion
as.data.table.data.table = function(x, key=NULL, ...) {
```
As mentioned above we don't need a keep.rownames argument for data.table. While you could theoretically set them, we don't even print them

Reply:
I understand the reasoning behind simplifying the signature. However, removing keep.rownames from as.data.table.data.table() creates an S3 method mismatch with as.data.table(), which formally declares ...(x, keep.rownames,...). This inconsistency leads to a 'formal argument' warning in R due to the differing signatures. While keep.rownames may not be commonly used with a data.table, it remains necessary to maintain correct dispatch consistency

Reply:
I haven't figured that out yet. I'll run the tests again to double-check whether the warning occurs as I described, and I'll get back to you with my findings

Reply:
I rechecked this, and the warning does occur as I described. Based on my research, the mismatch was previously masked by the use of ..., which prevented R’s checks from flagging it. Once keep.rownames was made explicit, R recognized the discrepancy between the S3 generic and method, as well as the documentation mismatch where the argument name and default values don’t align with the generic.

Reply:
It's probably [due to argument order](https://github.com/r-devel/r-svn/blob/2c9a39e7d042f31c468da8f82ab7a4edb475b8dd/src/library/tools/R/QC.R#L2675-L2680). The `as.data.table.data.table = function(x, ...)` method signature fits the `as.data.table = function(x, keep.rownames=FALSE, ...)` generic because `...` doesn't contradict the `keep.rownames` argument. On the other hand, `function(x, key, ...)` would contradict the generic signature because `key` goes first.

Reply:
That sounds right. Moreover, adding new arguments before `...` could constitute a breaking change for any downstreams relying on positional matching, c.f.

```r
as.data.table(data.table(a=1), '2')
```

Previously, the `'2'`is ignored, now it will be matched to `keep.rownames=`. Obviously pretty silly toy example, but it's more believable for this to happen accidentally in the case of more generic dispatch examples winding up with arguments in `...` perhaps unintentionally.

The upshot is -- new arguments should go after `...`. That requires them to be named, which I think is a good practice anyway, see e.g.

https://design.tidyverse.org/dots-after-required.html